### PR TITLE
Add cart_viewed tracks event (proof of concept)

### PR DESCRIPTION
**Pull request by haszari**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds a tracks event when the cart page is viewed. To allow this, various tracks code has been moved around.

- The PHP includes for core tracks facilities have moved from `WC_Admin` to `WooCommerce`.
- The entry point for tracking (`WC_Site_Tracking:init()`) is now triggered from `WC_Site_Tracking` constructor.
- There are two event prefixes defined in `WC_Tracks` - the existing `wcadmin_` and (proposed) `wcstore_` for front-end events.
- In `WC_Site_Tracking`, the JS tracks API is rendered on the front end and admin with the appropriate prefix baked in. 
- The front end tracks can be disabled with a filter - e.g. `add_filter( 'woocommerce_apply_frontend_user_tracking', __return_false );`

This is then used to trigger an example event from the cart page - see `WC_Cart_Tracking` for details.

In addition, I've added some default props (`url woo_version`) to all tracks events to do things like filter based on WooCommerce version. These props could be added in a separate PR; I've included them here because it's useful to have them on `cart_viewed`.

pb0Spc-ou-p2

### How to test the changes in this Pull Request:

1. Opt in to tracking in `WooCommerce > Settings > Advanced > WooCommerce.com`.
2. Perform various admin actions and check events are fired, with appropriate props. You can see events in Network tab - filter by `t.gif` and look at the query params.
3. Add something to cart and view cart page. Confirm a `cart_viewed` event is sent and that the props are all sensible.
4. Repeat tests with more interesting scenarios – anon user, logged in as customer, different browsers, etc.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

File: includes/tracks/class-wc-tracks.php [hunk 1]
```diff
@@ -11,9 +11,14 @@
...
-	 * Tracks event name prefix.
+	 * Event name prefix for admin tracks.
 	 */
-	const PREFIX = 'wcadmin_';
+	const ADMIN_PREFIX = 'wcadmin_';
+
+	/**
+	 * Event name prefix for front-end tracks.
+	 */
+	const FRONTEND_PREFIX = 'wcstore_';
```
Comment (psealock) [inline includes/tracks/class-wc-tracks.php]: Is this different from `woocommerceanalytics`?

Comment (haszari) [reply includes/tracks/class-wc-tracks.php]: Happy to change it to `woocommerceanalytics`. That seemed long and unclear to me, and not consistent with `wcadmin_` :)

**Review by psealock — COMMENTED** [review]
Is this cart tracking for c2? If so, it seems like you're doubling up the logic unless I'm seeing this wrong. ie, there is `window.wcTracks.recordEvent` and `_wca.push()`

File: includes/tracks/class-wc-site-tracking.php [hunk 2]
```diff
@@ -134,4 +156,48 @@ public static function init() {
 			}
 		}
 	}
+
+	/**
+	 * Set up frontend tracking.
+	 */
+	public static function init_frontend_events() {
```
Comment (deltaWhiskey) [inline includes/tracks/class-wc-site-tracking.php]: If we add an action to this function, plugins would be able to hook into this if they need to track front end stuff.

**Review by deltaWhiskey — APPROVED** [review]
Code looks good. I did not test. I recommend adding a `do_action()` to `init_frontend_events`.

With checkout conversion tracking going into WooCommerce core, it would be ideal if those stats were available to the store owner. For now, though, it makes sense that this builds upon the existing admin tracking architecture.

File: includes/tracks/class-wc-site-tracking.php [hunk 3]
```diff
@@ -134,4 +156,48 @@ public static function init() {
...
+			return;
+		}
+
+		self::enqueue_scripts();
+
+		if ( is_admin() ) {
+			self::init_admin_events();
+		}
+
+		self::init_frontend_events();
```
[the remark this answers is not in the document]
Comment (haszari) [reply includes/tracks/class-wc-site-tracking.php]: Might be worth adding a condition here e.g. `! ajax` - will see what other reviews think.

File: includes/tracks/class-wc-site-tracking.php [hunk 4]
```diff
@@ -52,23 +60,48 @@ public static function enqueue_scripts() {
...
 	}
 
 	/**
-	 * Adds the tracking function to the admin footer.
+	 * Renders a tracking function with specified event prefix.
+	 *
+	 * @param string $event_name_prefix Prefix to add to all triggered events.
 	 */
-	public static function add_tracking_function() {
+	public static function render_tracking_function( $event_name_prefix ) {
```
[the remark this answers is not in the document]
Comment (haszari) [reply includes/tracks/class-wc-site-tracking.php]: Thanks for the heads up - will keep an eye on it! cc @timmyc

File: includes/tracks/class-wc-site-tracking.php (hunk 2 above)
Comment (haszari) [reply includes/tracks/class-wc-site-tracking.php]: That's a great idea, I'll add that now. We'd likely use this this in WooCommerce Blocks plugin.

Comment (haszari) [reply includes/tracks/class-wc-site-tracking.php]: Added here: 17ec0809fb6e1ff5b8d12128f71d02025e457fa5

**Review by psealock — COMMENTED** [review]
Its my opinion that added C2 (customer's customers) tracking logic should live in the same repo as other code that performs the same function. The current code lives in Jetpack using the prefix `woocommerceanalytics`.

If we merge this code, anyone looking into a C2 event will need to poke around in Jetpack, Woocommerce and possibly wc-admin which houses the `window.wcTracks.recordEvent` logic. If a change in legal logic or permissions arises, it will need to be dealt with in two repos with separate release cycles. Users may have one update, but not the other. If you would like to add properties to all events in the future, you'll do in two repos. How can you trust the data while being unsure of how users are updating their plugins?

The solution will be, and should be, to migrate Woo Tracks events out of Jetpack and into Woo Core. Woocommerce will then fully own the code, its release cycles, handling of permissions, legal checks, and data will be more uniform and trustworthy.

**Comment by haszari** [comment]
For now, this PR is blocked until we decide to proceed with shopper/c2 events in core.

**Comment by haszari** [comment]
Closing this PR - we've decided to use [WooCommerce Analytics](https://jetpack.com/support/woocommerce-analytics/) for tracking performance/adoption of new cart & checkout blocks.

This PR still is a viable approach which might be worth switching to in future – i.e. support front-end/shopper tracking in WooCommerce core (including appropriate disclosure documentation and control for site builders/merchants, of course). However since there are no plans to ship this now I'm happy to close.

The `woo_version` prop on tracks events could be a useful addition in a separate PR – someone can add that when it comes up as a need.

**Pull request closed without merging**